### PR TITLE
[CodeCompletion] Don't provide 'init(nilLiteral:)' et al in optional context

### PR DESCRIPTION
**Pull request by rintaro**
For unresolved member completion, member decls in `Optional` referenceable with implicit member expression are useless in most cases.
- `init(_: <#Wrapped#>)`
- `init(nilLiteral: ())`
Instead, provide `nil` keyword with erasing `.` instruction.

As for `.some(<#Wrapped#>)` and `.none`, although they aren't useful in expression context, it's common to use them in pattern context. So don't hide them until we implement a mechanism to distinguish them.

rdar://problem/47806831

### CODE REVIEW

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -3634,6 +3637,14 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
           return false;
       }
 
+      // In optional context, ignore '.init(<some>)', 'init(nilLiteral:)',
+      // '.some(<some>)' and '.none'. They are useless in most cases.
+      if (T->getOptionalObjectType() &&
+          VD->getModuleContext()->isStdlibModule() &&
+          (isa<EnumElementDecl>(VD) || isa<ConstructorDecl>(VD))) {
```
Comment (jrose-apple) [inline lib/IDE/CodeCompletion.cpp]: Ah, you probably don't want to ignore *all* constructors. Someone could have added their own.

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: I think checking `VD->getModuleContext()->isStdlibModule()` is enough to limit them to `init(nilLiteral: ())` and `init(<#Wrapped#>)`. Added test case for that.
Is there any cases which hide user own initializers?

Comment (jrose-apple) [reply lib/IDE/CodeCompletion.cpp]: That should be good!

Comment (jrose-apple) [reply lib/IDE/CodeCompletion.cpp]: OTOH, maybe we should "just" have a `hideFromCodeCompletion` attribute…

**Comment by benlangmuir** [comment]
> init(_: <#Wrapped#>)
> init(nilLiteral: ())

Okay

> .some(<#Wrapped#>)
> .none

What's the justification for hiding these?  I've seen code that uses this style, particularly in switch cases.  It's more discoverable than "?" in a pattern binding.  Users from other languages with Optional may also be used to writing things this way.

**Comment by rintaro** [comment]
Hm, I was thinking mostly about expression context. But yes, `.some()`/`.none` is commonly used in matching pattern and I agree we should not hide them in pattern context.
I think, eventually, we should distinguish between pattern and expression context.

**Comment by rintaro** [comment]
Restored `.some` and `.none` for now with `TODO` note.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by benlangmuir — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**